### PR TITLE
Add a section on the git history to git.rst

### PR DESCRIPTION
**Pull request by jlaehne**
### Description of the change
Add documentation on keeping a sensible git history.

### Progress of the PR
- [x] update dev guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] ready for review.

### CODE REVIEW

**Comment by jlaehne** [comment]
@ericpre any suggestions, better links, ...? You were I think the first one who has started to clean up your git history regularly.

File: doc/dev_guide/git.rst [hunk 1]
```diff
@@ -102,3 +102,34 @@ For example, to rebase the ``bug_fix_branch`` branch from ``RELEASE_next_minor``
 
    $ git rebase --onto RELEASE_next_patch RELEASE_next_minor bug_fix_branch
 
+
+Cleaning the commit history
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (ericpre) [inline doc/dev_guide/git.rst]:
```suggestion
Keeping the git history clean
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: doc/dev_guide/git.rst [hunk 2]
```diff
@@ -102,3 +102,34 @@ For example, to rebase the ``bug_fix_branch`` branch from ``RELEASE_next_minor``
...
+
+For review, and for revisiting changes at a later point, it is advisable to clean the git history when a pull request is split up in too many commits. Commits that only fix formatting issues or failing tests usually can be *squashed* (merged) into the previous commits. When using a GUI for interaction with *git*, check out its features for joining and reordering commits.
```
Comment (ericpre) [inline doc/dev_guide/git.rst]:
```suggestion
For review, and for revisiting changes at a later point, it is advisable to keep a "clean" git history, i.e. a meaningful succession of commits. In some cases, it is useful to rewrite the git history to keep it more readable:
- it is not always possible to keep a clean history and quite often the code development follows an exploratory process with code changes going back and forth, etc. 
- Commits that only fix typographic mistakes, formatting or failing tests usually can be *squashed* (merged) into the previous commits.

When using a GUI for interaction with *git*, check out its features for joining and reordering commits.
```

File: doc/dev_guide/git.rst [hunk 3]
```diff
@@ -102,3 +102,34 @@ For example, to rebase the ``bug_fix_branch`` branch from ``RELEASE_next_minor``
...
+
+Afterwards you get a chance to edit the commit messages.
+
+Finally, to push the changes use a ``+`` in front of the branch name, to override commits you have already pushed to github previously:
+
+.. code:: bash
+
+  git push origin +lumberjack-branch
+
+See for example `Git good - The magic of keeping a clean git history <https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/>`_ for a more detailed blog post on this subject.
```
Comment (ericpre) [inline doc/dev_guide/git.rst]: This links doesn't have https and is flagged as insecure; maybe uses this link instead:
https://about.gitlab.com/blog/2018/06/07/keeping-git-commit-history-clean/?

**Review by ericpre — APPROVED** [review]
Thank you @jlaehne, this is good! I suggested some rewording and a different link.

File: doc/dev_guide/git.rst (hunk 3 above)
Comment (jlaehne) [reply doc/dev_guide/git.rst]: Very nice, it is more concise and includes some examples as well.

Comment (jlaehne) [inline doc/dev_guide/git.rst]:
```suggestion
See for example `How (and why!) to keep your Git commit history clean <https://about.gitlab.com/blog/2018/06/07/keeping-git-commit-history-clean/>`_ for a more detailed blog post on this subject.
```

**Pull request merged**